MacOS SDK Root issues with Vagrant 2.2.8

**Problem**
https://github.com/hashicorp/vagrant-installers/commit/c32d730a186bb5f7fec198231552f21ae0d4b60f updates the SDK root to be the static path `/Users/vagrant/SDKs/MacOSX10.9.sdk`, which is not present on client systems. Unfortunately, this change prevents plugins from being compiled as it uses the nonexistent path. Below is the log from installing `nokogiri` on MacOS 10.14.6.

```
$ vagrant --version                                     
Vagrant 2.2.8
$ vagrant plugin install nokogiri
Installing the 'nokogiri' plugin. This can take a few minutes...
Vagrant failed to properly resolve required dependencies. These
errors can commonly be caused by misconfigured plugin installations
or transient network issues. The reported error is:

ERROR: Failed to build gem native extension.

    current directory: /Users/username/.vagrant.d/gems/2.6.6/gems/nokogiri-1.10.9/ext/nokogiri
/opt/vagrant/embedded/bin/ruby -I /opt/vagrant/embedded/lib/ruby/2.6.0 -r ./siteconf20200505-85900-1a75xhz.rb extconf.rb
checking if the C compiler accepts -I/opt/vagrant/embedded/include -I/opt/vagrant/embedded/include/libxml2 -I /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/libxml2... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

...

extconf failed, exit code 1

```

The `mkmf.log` shows the path: `/Users/vagrant/SDKs/MacOSX10.9.sdk`

```
"clang -o conftest -I/opt/vagrant/embedded/include/ruby-2.6.0/x86_64-darwin19 -I/opt/vagrant/embedded/include/ruby-2.6.0/ruby/backward -I/opt/vagrant/embedded/include/ruby-2.6.0 -I. -I/opt/vagrant/embedded/include -mmacosx-version-min=10.9 -isysroot /Users/vagrant/SDKs/MacOSX10.9.sdk -I/opt/vagrant/embedded/include -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -I/opt/vagrant/embedded/include -mmacosx-version-min=10.9 -isysroot /Users/vagrant/SDKs/MacOSX10.9.sdk -fno-common -pipe -I/opt/vagrant/embedded/include -I/opt/vagrant/embedded/include/libxml2 -I /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/libxml2 conftest.c  -L. -L/opt/vagrant/embedded/lib -L/opt/vagrant/embedded/lib -L.  -mmacosx-version-min=10.9 -isysroot /Users/vagrant/SDKs/MacOSX10.9.sdk -fstack-protector-strong -L/opt/vagrant/embedded/lib     -lruby.2.6    "
clang: warning: no such sysroot directory: '/Users/vagrant/SDKs/MacOSX10.9.sdk' [-Wmissing-sysroot]
In file included from conftest.c:1:
In file included from /opt/vagrant/embedded/include/ruby-2.6.0/ruby.h:33:
In file included from /opt/vagrant/embedded/include/ruby-2.6.0/ruby/ruby.h:29:
/opt/vagrant/embedded/include/ruby-2.6.0/ruby/defines.h:123:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return 0;
6: }
/* end */

```

**Workaround**
Updating the path to `/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk` allows for the gems to be installed.

```
$ sudo gsed -i 's#/Users/vagrant/SDKs/MacOSX10.9.sdk#/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk#'  /opt/vagrant/embedded/bin/curl-config /opt/vagrant/embedded/lib/pkgconfig/ruby-2.6.pc /opt/vagrant/embedded/lib/ruby/2.6.0/x86_64-darwin19/rbconfig.rb
$ vagrant plugin install nokogiri
Installing the 'nokogiri' plugin. This can take a few minutes...
Fetching mini_portile2-2.4.0.gem
Fetching nokogiri-1.10.9.gem
Building native extensions. This could take a while...
Installed the plugin 'nokogiri (1.10.9)'!
```

On my system `/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk` is a symlink to the version specific path `/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk`, which is what I get from `xcrun --sdk macosx --show-sdk-path`

--- a/package/support/package_darwin.sh
+++ b/package/support/package_darwin.sh
@@ -7,14 +7,27 @@ if [ "$#" -ne "2" ]; then
   exit 1
 fi
 
-export MACOSX_DEPLOYMENT_TARGET="10.9"
-export SDKROOT="/Users/vagrant/SDKs/MacOSX10.9.sdk" #"$(xcrun --sdk macosx --show-sdk-path)"
+macos_deployment_target="10.9"
+
+sdk_root="/Library/Developer/CommandLineTools/SDKs"
+sdk_path="${sdk_root}/MacOSX.sdk"
+versioned_sdk_path="${sdk_root}/MacOSX${macos_deployment_target}.sdk"
+# Check that deployment target sdk exists
+if [ ! -d "${versioned_sdk_path}" ]; then
+    echo_stderr " !! Requested macOS SDK version is not available: ${macos_deployment_target}"
+    exit 1
+else
+    rm -f "${sdk_path}"
+    ln -s "${versioned_sdk_path}" "${sdk_path}"
+fi
+export MACOSX_DEPLOYMENT_TARGET="${macos_deployment_target}"
+export SDKROOT="${sdk_path}" #"$(xcrun --sdk macosx --show-sdk-path)"
 export ISYSROOT="-isysroot ${SDKROOT}"
 export SYSLIBROOT="-syslibroot ${SDKROOT}"
 export SYS_ROOT="${SDKROOT}"
-export CFLAGS="-mmacosx-version-min=10.9 ${ISYSROOT}"
+export CFLAGS="-mmacosx-version-min=${macos_deployment_target} ${ISYSROOT}"
 export CXXFLAGS="${CFLAGS}"
-export LDFLAGS="-mmacosx-version-min=10.9 ${SYSLIBROOT}"
+export LDFLAGS="-mmacosx-version-min=${macos_deployment_target} ${SYSLIBROOT}"
 
 # Get our directory
 SOURCE="${BASH_SOURCE[0]}"

--- a/package/vagrant-scripts/osx.sh
+++ b/package/vagrant-scripts/osx.sh
@@ -11,6 +11,12 @@ mv gon /System/Volumes/Data/usr/local/bin/gon
 
 su vagrant -l -c 'brew install bash'
 
+# Move the SDK into the developer section
+sdk="/Users/vagrant/SDKs/MacOSX10.9.sdk"
+if [ -d "${sdk}" ]; then
+    mv "${sdk}" /Library/Developer/CommandLineTools/SDKs/
+fi
+
 chmod 755 /vagrant/package/package.sh
 
 set -e

--- a/substrate/run.sh
+++ b/substrate/run.sh
@@ -20,6 +20,8 @@ ruby_version="2.6.6"
 xz_version="5.2.4"
 zlib_version="1.2.11"
 
+macos_deployment_target="10.9"
+
 function echo_stderr {
     (>&2 echo "$@")
 }
@@ -191,18 +193,28 @@ export CPPFLAGS="-I${embed_dir}/include"
 export LDFLAGS="-L${embed_dir}/lib"
 ORIGINAL_LDFLAGS="${LDFLAGS}"
 if [[ "${host_os}" = "darwin" ]]; then
-    export MACOSX_DEPLOYMENT_TARGET="10.9"
-    export SDKROOT="/Users/vagrant/SDKs/MacOSX10.9.sdk" #"$(xcrun --sdk macosx --show-sdk-path)"
+    sdk_root="/Library/Developer/CommandLineTools/SDKs"
+    sdk_path="${sdk_root}/MacOSX.sdk"
+    versioned_sdk_path="${sdk_root}/MacOSX${macos_deployment_target}.sdk"
+    # Check that deployment target sdk exists
+    if [ ! -d "${versioned_sdk_path}" ]; then
+        echo_stderr " !! Requested macOS SDK version is not available: ${macos_deployment_target}"
+        exit 1
+    else
+        rm -f "${sdk_path}"
+        ln -s "${versioned_sdk_path}" "${sdk_path}"
+    fi
+    export MACOSX_DEPLOYMENT_TARGET="${macos_deployment_target}"
+    export SDKROOT="${sdk_path}" #"$(xcrun --sdk macosx --show-sdk-path)"
     export ISYSROOT="-isysroot ${SDKROOT}"
     export SYSLIBROOT="-syslibroot ${SDKROOT}"
     export SYS_ROOT="${SDKROOT}"
-    export CFLAGS="${CFLAGS} -mmacosx-version-min=10.9 ${ISYSROOT}"
+    export CFLAGS="${CFLAGS} -mmacosx-version-min=${macos_deployment_target} ${ISYSROOT}"
     export CXXFLAGS="${CFLAGS}"
     export CPPFLAGS="${CFLAGS}"
-    export LDFLAGS="${LD_FLAGS} -mmacosx-version-min=10.9 ${ISYSROOT}" # ${SYSLIBROOT}"
+    export LDFLAGS="${LD_FLAGS} -mmacosx-version-min=${macos_deployment_target} ${ISYSROOT}" # ${SYSLIBROOT}"
 
     export LD_RPATH="/opt/vagrant/embedded/lib:/opt/vagrant/embedded/lib64"
-    ORIGINAL_LD_RPATH="${LD_RPATH}"
     libtool="glibtool"
 else
     export LDFLAGS="${LDFLAGS} -L${embed_dir}/lib64 -Wl,-rpath=/opt/vagrant/embedded/lib:/opt/vagrant/embedded/lib64"
@@ -368,18 +380,14 @@ if [ $? -ne 0 ]; then
 fi
 tar -xzf openssl.tar.gz
 pushd openssl-*
-if [ "${LD_RPATH}" != "" ]; then
-    export LD_RPATH="/opt/vagrant/embedded/lib:/opt/vagrant/embedded/lib64"
-else
+if [ -z "${LD_RPATH}" ]; then
     CURRENT_LDFLAGS="${LDFLAGS}"
     export LDFLAGS="${ORIGINAL_LDFLAGS} -Wl,-rpath=/opt/vagrant/embedded/lib"
 fi
 ./config --prefix="${embed_dir}" --openssldir="${embed_dir}" shared
 make
 make install_sw
-if [ "${LD_RPATH}" != "" ]; then
-    export LD_RPATH="${ORIGINAL_LD_RPATH}"
-else
+if [ -z "${LD_RPATH}" ]; then
     export LDFLAGS="${CURRENT_LDFLAGS}"
 fi
 popd

--- a/substrate/vagrant-scripts/osx.sh
+++ b/substrate/vagrant-scripts/osx.sh
@@ -18,6 +18,12 @@ mv cacert.pem /usr/local/etc/openssl/cacert.pem
 export SSL_CERT_FILE=/usr/local/etc/openssl/cacert.pem
 export PATH=$PATH:/usr/local/bin:/usr/local/go/bin
 
+# Move the SDK into the developer section
+sdk="/Users/vagrant/SDKs/MacOSX10.9.sdk"
+if [ -d "${sdk}" ]; then
+    mv "${sdk}" /Library/Developer/CommandLineTools/SDKs/
+fi
+
 set -e
 
 if [ "${VAGRANT_BUILD_DEBUG}" = "1" ]; then

